extensions/oauth-integration-validator/app.R

```diff
@@ -0,0 +1,122 @@
+library(shiny)
+library(httr2)
+library(bslib)
+library(connectapi)
+
+ui <- page_sidebar(
+  title = "OAuth Integration Validation in R",
+  sidebar = sidebar(
+    title = "Request an endpoint",
+    textInput("endpoint", "GET endpoint to request:", placeholder = "https://example.com/api/list"),
+    actionButton("request", "Request")
+  ),
+  layout_columns(
+    card(
+      card_header("Results"),
+      htmlOutput("results")
+    )
+  )
+)
+
+server <- function(input, output, session) {
+  
+  # check if running on Posit Connect
+  if (Sys.getenv("RSTUDIO_PRODUCT") == "CONNECT") {
+    # initialize Connect API client
+    client <- connect()
+    
+    # read the user-session-token header
+    user_session_token <- session$request$HTTP_POSIT_CONNECT_USER_SESSION_TOKEN
+    
+    # wrap client credentials exchange request for error handling
+    # and so the app does not fail to start on first deploy
+    credentials <- reactive({
+      tryCatch({
+        # exchange user session token for an access token using connectapi client
+        # checking for a client error in
+        credentials <- get_oauth_credentials(client, user_session_token)
+      }, error = function(e) {
+          return(list(error = "Unable to obtain access token: ", message = e$message))
+        })
+      })
+   
+    # set token to null
+    token <- reactiveVal(NULL)
+    
+    # react to the credentials request
+    observeEvent(credentials(), {
+      creds <- credentials()
+      
+      # check if the credentials exchange was successful
+      if (is.null(creds$error)) {
+        token(creds$access_token)
+      } else {
+        # show notification to user
+        showNotification(creds$error, creds$message)
+      }
+    }) 
+  } else {
+    # grab the access token from the OAUTH_TOKEN environment variable if running
+    # locally
+    token <- Sys.getenv("OAUTH_TOKEN")
+  }
+  
+  # react to request button being pressed 
+  observeEvent(input$request, {
+    # do not allow empty value for endpoint
+    req(input$endpoint)
+    endpoint <- input$endpoint
+    
+    # do not allow request without an access token
+    if (is.null(token())) {
+      # remind the user that they don't have a token 
+      output$results <- renderUI({
+        HTML('<span style="color:red; font-size:larger;">No access token found.',
+        '<br/>Try logging out and logging back in to the integration if you have not ',
+        'visited this content recently.<br/>If an integration is not associated with ',
+        'the content then ask your publisher to add one.</span><br/>')
```

Comment:
Nice! I like the interactive debugging message. :)

I was initially going to suggest putting the bit about checking that there is an extension associated with the content first, but I guess if the app is targeted towards people who aren't viewing it in the Dashboard, this might be less relevant.

Perhaps "the content owner" instead of "your publisher" on the last line? Soft suggestion.